Fixes Random Desyncs Due To Statue Exception And NextThingId Missmatch

Tick-git opened the pull request:
Label: 1.6, High Priority, Fix

- This problem causes save files to constantly desync.
- The issue is caused by the new 1.6 statue created in the art bench.
- The root cause happens during the statue expose process, a nextThingId mismatch occurs because the host uses negative local thing IDs (patched by us) while the clients use unpatched ones.
- I suspect the problem occurs more often when flying with the gravship, since a statue can sometimes be present on NPC maps (just a guess).

### Tested:
- Test that damaged save files work correctly
- Test creating a Statue with art bench
- Spawning a Statue via debug tool (Statues look different but do not lead to problems)

### This PR fixes two issues with the new method CompStatue.InitFakePawn:

#

1. **Pawn name sync issue:** The pawn getter is synced while pawn is not fully initialized, which caused syncing errors due to a missing PawnKindDef.

#

2. **Desyncs due to UniqueIdsPatch (nextThingId missmatch):** This was the main reason I investigated the issue. Many desyncs occurred because UniqueIdsPatch (look link below for code) generates negative thing IDs if certain values are set in a specific way. These values differ between client and host - specifically the Multiplayer.InInterface property. The method InitFakePawn is called inside an ExecuteWhenFinished event (look below for code). On the host, this event runs as an async long event, while on the client it runs as a synced event. As a result, the LongEventHandler.currentEvent field is different on host and client, which makes InInterface differ as well. This causes the host to generate negative new thing IDs, while the clients generate positive ones. Since clients get ahead in nextThingIds, the desyncs occur

<img width="1920" height="415" alt="image" src="https://github.com/user-attachments/assets/4de45e08-6ceb-4bb6-84be-a8d58a2e5d88" />

```
class CompStatue
...
public override void PostExposeData()
{
    ...
    if (Scribe.mode == LoadSaveMode.PostLoadInit)
    {
        ...
        LongEventHandler.ExecuteWhenFinished(new Action(this.InitFakePawn));
    }
}
```

https://github.com/rwmt/Multiplayer/blob/8b9b44e3e2040302565c3a17b581e65b2e66019a/Source/Client/Patches/UniqueIds.cs#L9C5-L31C6

#

### Important:
I'm not entirely sure if the fix is the right approach. I haven’t looked into why we use two different events or why we rely on local IDs that are negative values. If the fix feels off to you, I can dig deeper into the problem.

SokyranTheDragon commented:
> I'm not entirely sure if the fix is the right approach. I haven’t looked into why we use two different events **or why we rely on local IDs that are negative values.** If the fix feels off to you, I can dig deeper into the problem.

To explain the highlighted part:

The game expects a lot of objects in the game to have a unique ID, and desyncs will happen when they are out of sync between players.

However, sometimes we don't want the IDs to be in syncs. Usually in situations when something happens only for one of the clients - for example, displaying a message when rejecting input. It would be pointless to display a message for every player when it's related to something one person did. Likewise, syncing a message saying "increment IDs by 1" every time such a message is generated would be wasteful - especially since those messages can get quite spammy.

So a simple solution to that is to just use negative IDs. Anything with a negative ID is considered a local object that is not synced with all players (unless a bug such as this one occurs). The game never really uses negative values for ID (with exception of -1/-2/-3), meaning it's pretty safe to do so as well. And when we're syncing something using local IDs with other players - that's when such object gets assigned a positive ID.

SokyranTheDragon commented:
I believe that modifying `UpdateCurrentAsynchronousEvent` _should_ be enough to fix this issue (and other potential ones, likely related to mods).

However, if this is not the case, then we'd need to look into other solutions first.

Tick-git commented:
Thanks for the feedback. I was also looking at the cleanup order, but I assumed there was an intentional reason behind the sequence.

Would it be okay if I just inject a method via the transpiler for LongEventHandler.ExecuteToExecuteWhenFinished that simply does something like this:

```
injectedMethod
    LongEventHandler.currentEvent = null;
    LongEventHandler.eventThread = null;
    LongEventHandler.levelLoadOp = null;
    LongEventHandler.ExecuteToExecuteWhenFinished()
```
We'll clean it up twice this way, but the transpiler would stay small.

#

> Pawn name sync issue: The pawn getter is synced while pawn is not fully initialized, which caused syncing errors due to a missing PawnKindDef.

I think this will still be a problem though. Looking at it again tomorrow.

Tick-git commented:
I have already written a comment and deleted it due me looking wrong at the code and making a wrong suggestion. Looking into it tomorrow again - Thanks for the feedback :)

Tick-git commented:
I implemented the suggested changes for Problem 2 and updated the logic for Problem 1 by replacing `pawn.Name` with `pawn.nameInt`.

I created two variations of your proposal:

* **Option A** (included in this PR): uses a simple transpiler to change the cleanup logic for both **Multiplayer** and **Singleplayer**.
* **Option B**: applies the same logic, but only in **Multiplayer**.

You can see the implementation for Option B here. I kept the logs so you can review how I tested the functionality, since this is the first time I’ve done it this way:

https://github.com/Tick-git/Multiplayer/blob/95904c636c01da663765ebdd63fcb79868929251/Source/Client/Patches/LongEvents.cs#L86-L141

Tick-git commented:
I reverted the change and opened https://github.com/rwmt/Multiplayer/issues/654 so the additional `InInterface` condition can be added later

notfood reviewed (approved):
Looks good, ready?
